allennlp/data/transforms.py

```diff
@@ -0,0 +1,558 @@
+from typing import Dict, Tuple, List, Iterable, Generic, TypeVar, Deque, Union, Any
+import itertools
+from collections import deque, defaultdict
+import random
+
+import torch
+from torch.utils.data import IterableDataset as IterableTorchDataset
+
+
+from allennlp.common.registrable import Registrable
+from allennlp.data.instance import Instance
+from allennlp.data.vocabulary import Vocabulary
+
+from allennlp.data.fields import MetadataField
+
+from allennlp.common.util import lazy_groups_of
+
+
+Batched = Iterable[Instance]
+InstanceOrBatch = TypeVar("InstanceOrBatch", Iterable[Instance], Instance)
+
+
+class DatasetFromGenerator(IterableTorchDataset):
+    def __init__(self, generator: Iterable[Union[Iterable[Instance], Instance]]):
+        self.generator = generator
+
+    def __iter__(self) -> Iterable[Union[Iterable[Instance], Instance]]:
+
+        for x in self.generator:
+            yield x
+
+
+class Transform(IterableTorchDataset, Registrable):
```

Comment:
Maybe this is due to my background in other languages, but I'd like to push back a bit on playing loose with types. Though mypy doesn't always do the best job of enforcing type constraints, the type annotations are still very useful for communicating with the reader of the code. I've been tripped up in Python due to type mismatches and I think that sort of hack should be reserved for extraordinary cases.

Reply:
Using `Any` as a type annotation doesn't get you any type safety, anyway, so I'm really not sure what you gain with this separation. It seems strictly better to just have one type, `Transform`, and parameterizing both its input and output if you really care.

The only thing I'm really complaining about is the external use of `InstanceTransform`. If the external type annotations all use `List[Transform]`, I'd be fine with this. I basically suggested doing it in another comment. But I'd still make `Transform` generic, instead of using Any.

Reply:
+1 for generics over `Any`!